# Exercise 02 -  OLAP Cubes - Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by creating and connecting to the database by running the cells below.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Start with a simple cube
TODO: Write a query that calculates the revenue (sales_amount) by day, rating, and city. Remember to join with the appropriate dimension tables to replace the keys with the dimension labels. Sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below.

In [5]:
%%time
%%sql

SELECT d.day, m.rating, s.city, sum(f.sales_amount) as total_sales
FROM factsales f
join dimmovie m on m.movie_key = f.movie_key
join dimstore s on s.store_key = f.store_key
join dimdate d on d.date_key = f.date_key
group by d.day, m.rating, s.city

 * postgresql://student:***@127.0.0.1:5432/pagila_star
279 rows affected.
CPU times: user 1.22 ms, sys: 4.16 ms, total: 5.38 ms
Wall time: 29 ms


[(14, 'R', 'Lethbridge', Decimal('66.78')),
 (18, 'NC-17', 'Woodridge', Decimal('443.99')),
 (20, 'G', 'Lethbridge', Decimal('347.11')),
 (27, 'PG-13', 'Woodridge', Decimal('444.90')),
 (23, 'G', 'Lethbridge', Decimal('183.56')),
 (30, 'PG', 'Woodridge', Decimal('593.57')),
 (26, 'NC-17', 'Woodridge', Decimal('156.68')),
 (6, 'R', 'Woodridge', Decimal('209.52')),
 (31, 'G', 'Lethbridge', Decimal('57.89')),
 (7, 'PG', 'Lethbridge', Decimal('204.50')),
 (17, 'R', 'Lethbridge', Decimal('407.04')),
 (21, 'R', 'Woodridge', Decimal('404.06')),
 (28, 'R', 'Lethbridge', Decimal('347.17')),
 (2, 'R', 'Woodridge', Decimal('250.40')),
 (29, 'R', 'Woodridge', Decimal('363.17')),
 (15, 'R', 'Lethbridge', Decimal('91.72')),
 (19, 'PG-13', 'Lethbridge', Decimal('465.87')),
 (5, 'PG', 'Lethbridge', Decimal('18.96')),
 (20, 'PG-13', 'Lethbridge', Decimal('538.93')),
 (14, 'PG', 'Woodridge', Decimal('69.78')),
 (25, 'NC-17', 'Woodridge', Decimal('51.87')),
 (31, 'PG-13', 'Lethbridge', Decimal('57.86')),
 (30, 'R', 'Lethbridge', Decimal('683.46')),
 (27, 'G', 'Woodridge', Decimal('313.33')),
 (23, 'PG-13', 'Lethbridge', Decimal('266.37')),
 (6, 'PG', 'Lethbridge', Decimal('203.54')),
 (7, 'R', 'Woodridge', Decimal('183.58')),
 (1, 'NC-17', 'Woodridge', Decimal('281.28')),
 (28, 'PG', 'Woodridge', Decimal('378.12')),
 (2, 'PG', 'Lethbridge', Decimal('197.48')),
 (21, 'PG', 'Lethbridge', Decimal('348.12')),
 (17, 'PG', 'Woodridge', Decimal('400.01')),
 (10, 'NC-17', 'Woodridge', Decimal('184.53')),
 (29, 'PG', 'Lethbridge', Decimal('377.09')),
 (12, 'NC-17', 'Woodridge', Decimal('218.51')),
 (19, 'G', 'Lethbridge', Decimal('383.06')),
 (5, 'R', 'Woodridge', Decimal('26.92')),
 (15, 'PG', 'Woodridge', Decimal('138.66')),
 (31, 'PG', 'Woodridge', Decimal('62.87')),
 (27, 'R', 'Woodridge', Decimal('336.14')),
 (30, 'G', 'Lethbridge', Decimal('730.48')),
 (23, 'PG', 'Woodridge', Decimal('242.45')),
 (6, 'PG-13', 'Woodridge', Decimal('191.47')),
 (11, 'NC-17', 'Lethbridge', Decimal('217.52')),
 (20, 'PG', 'Woodridge', Decimal('376.12')),
 (14, 'PG-13', 'Lethbridge', Decimal('83.76')),
 (29, 'PG-13', 'Woodridge', Decimal('326.18')),
 (9, 'NC-17', 'Woodridge', Decimal('256.37')),
 (5, 'G', 'Woodridge', Decimal('26.95')),
 (19, 'R', 'Lethbridge', Decimal('408.02')),
 (15, 'PG-13', 'Lethbridge', Decimal('151.61')),
 (24, 'NC-17', 'Woodridge', Decimal('11.98')),
 (28, 'PG-13', 'Lethbridge', Decimal('455.97')),
 (22, 'NC-17', 'Woodridge', Decimal('262.31')),
 (2, 'PG-13', 'Woodridge', Decimal('290.34')),
 (21, 'PG-13', 'Woodridge', Decimal('391.05')),
 (17, 'PG-13', 'Lethbridge', Decimal('368.09')),
 (16, 'NC-17', 'Lethbridge', Decimal('136.66')),
 (7, 'G', 'Woodridge', Decimal('164.58')),
 (27, 'PG', 'Lethbridge', Decimal('288.27')),
 (6, 'G', 'Woodridge', Decimal('156.69')),
 (30, 'PG-13', 'Lethbridge', Decimal('784.21')),
 (31, 'R', 'Lethbridge', Decimal('45.87')),
 (23, 'R', 'Lethbridge', Decimal('277.40')),
 (20, 'R', 'Lethbridge', Decimal('388.18')),
 (14, 'G', 'Lethbridge', Decimal('64.79')),
 (8, 'NC-17', 'Lethbridge', Decimal('255.47')),
 (15, 'G', 'Lethbridge', Decimal('110.73')),
 (29, 'G', 'Woodridge', Decimal('282.30')),
 (19, 'PG', 'Woodridge', Decimal('423.93')),
 (5, 'PG-13', 'Woodridge', Decimal('36.92')),
 (21, 'G', 'Woodridge', Decimal('373.20')),
 (28, 'G', 'Lethbridge', Decimal('275.26')),
 (2, 'G', 'Woodridge', Decimal('234.46')),
 (17, 'G', 'Lethbridge', Decimal('306.30')),
 (7, 'PG-13', 'Woodridge', Decimal('178.61')),
 (23, 'NC-17', 'Lethbridge', Decimal('208.50')),
 (11, 'PG', 'Woodridge', Decimal('158.58')),
 (26, 'G', 'Woodridge', Decimal('76.84')),
 (31, 'NC-17', 'Lethbridge', Decimal('70.84')),
 (25, 'PG-13', 'Woodridge', Decimal('68.84')),
 (18, 'G', 'Woodridge', Decimal('356.17')),
 (20, 'NC-17', 'Lethbridge', Decimal('374.12')),
 (24, 'PG', 'Lethbridge', Decimal('9.99')),
 (12, 'PG-13', 'Woodridge', Decimal('244.44')),
 (9, 'PG', 'Lethbridge', Decimal('183.55')),
 (8, 'R', 'Lethbridge', Decimal('186.53')),
 (22, 'PG', 'Lethbr

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

TODO: Write a query that reduces the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below. 

In [6]:
%%time
%%sql

SELECT d.day, m.rating, s.city, sum(f.sales_amount) as total_sales
FROM factsales f
join dimmovie m on m.movie_key = f.movie_key
join dimstore s on s.store_key = f.store_key
join dimdate d on d.date_key = f.date_key
where m.rating = 'PG-13'
group by d.day, m.rating, s.city

 * postgresql://student:***@127.0.0.1:5432/pagila_star
56 rows affected.
CPU times: user 4.79 ms, sys: 225 µs, total: 5.02 ms
Wall time: 14.1 ms


[(14, 'PG-13', 'Woodridge', Decimal('64.76')),
 (27, 'PG-13', 'Woodridge', Decimal('444.90')),
 (25, 'PG-13', 'Woodridge', Decimal('68.84')),
 (6, 'PG-13', 'Lethbridge', Decimal('243.44')),
 (21, 'PG-13', 'Lethbridge', Decimal('499.92')),
 (12, 'PG-13', 'Woodridge', Decimal('244.44')),
 (28, 'PG-13', 'Woodridge', Decimal('368.12')),
 (2, 'PG-13', 'Lethbridge', Decimal('274.36')),
 (17, 'PG-13', 'Woodridge', Decimal('488.83')),
 (8, 'PG-13', 'Woodridge', Decimal('274.41')),
 (29, 'PG-13', 'Lethbridge', Decimal('414.09')),
 (10, 'PG-13', 'Woodridge', Decimal('208.44')),
 (15, 'PG-13', 'Woodridge', Decimal('195.54')),
 (19, 'PG-13', 'Lethbridge', Decimal('465.87')),
 (1, 'PG-13', 'Woodridge', Decimal('306.30')),
 (11, 'PG-13', 'Woodridge', Decimal('257.42')),
 (20, 'PG-13', 'Lethbridge', Decimal('538.93')),
 (26, 'PG-13', 'Woodridge', Decimal('101.76')),
 (30, 'PG-13', 'Woodridge', Decimal('635.48')),
 (18, 'PG-13', 'Woodridge', Decimal('428.93')),
 (31, 'PG-13', 'Lethbridge', Decimal('57.86')),
 (23, 'PG-13', 'Lethbridge', Decimal('266.37')),
 (24, 'PG-13', 'Lethbridge', Decimal('18.97')),
 (7, 'PG-13', 'Lethbridge', Decimal('228.48')),
 (9, 'PG-13', 'Lethbridge', Decimal('261.38')),
 (22, 'PG-13', 'Lethbridge', Decimal('347.18')),
 (16, 'PG-13', 'Woodridge', Decimal('169.60')),
 (5, 'PG-13', 'Lethbridge', Decimal('52.90')),
 (27, 'PG-13', 'Lethbridge', Decimal('374.14')),
 (25, 'PG-13', 'Lethbridge', Decimal('105.77')),
 (6, 'PG-13', 'Woodridge', Decimal('191.47')),
 (14, 'PG-13', 'Lethbridge', Decimal('83.76')),
 (29, 'PG-13', 'Woodridge', Decimal('326.18')),
 (10, 'PG-13', 'Lethbridge', Decimal('207.53')),
 (19, 'PG-13', 'Woodridge', Decimal('430.01')),
 (15, 'PG-13', 'Lethbridge', Decimal('151.61')),
 (1, 'PG-13', 'Lethbridge', Decimal('310.30')),
 (28, 'PG-13', 'Lethbridge', Decimal('455.97')),
 (2, 'PG-13', 'Woodridge', Decimal('290.34')),
 (21, 'PG-13', 'Woodridge', Decimal('391.05')),
 (12, 'PG-13', 'Lethbridge', Decimal('245.47')),
 (8, 'PG-13', 'Lethbridge', Decimal('283.34')),
 (17, 'PG-13', 'Lethbridge', Decimal('368.09')),
 (30, 'PG-13', 'Lethbridge', Decimal('784.21')),
 (18, 'PG-13', 'Lethbridge', Decimal('466.92')),
 (31, 'PG-13', 'Woodridge', Decimal('52.88')),
 (23, 'PG-13', 'Woodridge', Decimal('234.45')),
 (11, 'PG-13', 'Lethbridge', Decimal('272.39')),
 (20, 'PG-13', 'Woodridge', Decimal('362.00')),
 (26, 'PG-13', 'Lethbridge', Decimal('115.73')),
 (22, 'PG-13', 'Woodridge', Decimal('266.40')),
 (16, 'PG-13', 'Lethbridge', Decimal('196.53')),
 (5, 'PG-13', 'Woodridge', Decimal('36.92')),
 (7, 'PG-13', 'Woodridge', Decimal('178.61')),
 (24, 'PG-13', 'Woodridge', Decimal('4.98')),
 (9, 'PG-13', 'Woodridge', Decimal('144.59'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

TODO: Write a query to create a subcube of the initial cube that includes moves with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

The first few rows of your output should match the table below. 

In [7]:
%%time
%%sql

SELECT d.day, m.rating, s.city, sum(f.sales_amount) as total_sales
FROM factsales f
join dimmovie m on m.movie_key = f.movie_key
join dimstore s on s.store_key = f.store_key
join dimdate d on d.date_key = f.date_key
where m.rating in ('PG-13', 'PG') 
    and s.city in ('Bellevue','Lancaster')
    and d.day in (1, 15, 30)
group by d.day, m.rating, s.city

 * postgresql://student:***@127.0.0.1:5432/pagila_star
0 rows affected.
CPU times: user 5.93 ms, sys: 0 ns, total: 5.93 ms
Wall time: 8.7 ms


[]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>